### PR TITLE
optimization for dot(csr.T, dense) = rsp

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##

Use prefix sum to compute `nnr` in order to allocate the row_sparse output.

Currently `dot(csr.T, dense) = rsp` will allocate the dense output and then cast it to row_sparse, but not free the unused memory.

I use `run_benchmark(context, lhs="csr", rhs="default", lhs_trans=True, ...)` in `mxnet/benchmark/python/sparse/dot.py`. Please correct me if I'm wrong.

But is `dot(csr.T, dense) = rsp` in master slow like this? Might due to others are using my machine at the same time?

Performance of origin `dot(csr.T, dense) = rsp`,

```
[hanfeng@model-gpu00:sparse]$ python dot.py --num-omp-threads 20
========================================================
  mxnet sparse dot benchmark: dot(csr, default) = default
  (matrix multiplication: (m x k)^T * (k x n) = m x n)
========================================================
 lhs_density(%)  rhs_density(%)    context        m        k        n  t_sparse(ms)   t_dense(ms)  speedup
            1.0           100.0     cpu(0)      128  1000000      256        366.19        135.76     0.37
            1.0           100.0     cpu(0)      128  1000000     1000       1327.12        503.92     0.38
            1.0           100.0     cpu(0)      128  1000000     1000       1237.33        454.01     0.37
            1.0           100.0     cpu(0)       64  1000000     1000        868.38        345.38     0.40
            1.0           100.0     cpu(0)      128  1000000     1000       1237.09        437.32     0.35
```

After this PR,
```
[hanfeng@model-gpu00:sparse]$ python dot.py --num-omp-threads 20
========================================================
  mxnet sparse dot benchmark: dot(csr, default) = default
  (matrix multiplication: (m x k)^T * (k x n) = m x n)
========================================================
 lhs_density(%)  rhs_density(%)    context        m        k        n  t_sparse(ms)   t_dense(ms)  speedup
            1.0           100.0     cpu(0)      128  1000000      256         83.90        137.18     1.64
            1.0           100.0     cpu(0)      128  1000000     1000        410.63        448.30     1.09
            1.0           100.0     cpu(0)      128  1000000     1000        467.91        492.87     1.05
            1.0           100.0     cpu(0)       64  1000000     1000        259.99        348.32     1.34
            1.0           100.0     cpu(0)      128  1000000     1000        481.77        416.20     0.86
```
As a feature requested in https://github.com/apache/incubator-mxnet/issues/8168

cc @eric-haibin-lin 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] unittests already exist

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Intersting edge cases to note here

### CODE REVIEW

File: src/operator/tensor/dot-inl.h [hunk 1]
```diff
@@ -573,6 +577,18 @@ inline void DotCsrDnsDnsImpl(const OpContext& ctx,
   });
 }
 
+
+struct MarkCsrColKernel {
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: Can we reuse this kernel and give it a better name?
https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/util/tensor_util-inl.h#L37

File: src/operator/tensor/dot-inl.h [hunk 2]
```diff
@@ -598,40 +614,52 @@ inline void DotCsrDnsRspImpl(const OpContext& ctx,
...
-          mxnet_op::Kernel<set_zero, cpu>::Launch(s, num_threads, row_idx);
-          num_threads = mxnet_op::get_num_threads<cpu>(data_out.shape_[0]);
-          dim_t seg_len = (data_out.shape_[0] + num_threads - 1) / num_threads;
+          RType* row_idx_out = row_idx.dptr<RType>();
+
+          mxnet_op::Kernel<FillRspRowIdxKernel, cpu>::Launch(s, num_rows,
+            row_idx_out, prefix_sum, num_rows);
+
+          num_threads = mxnet_op::get_num_threads<cpu>(ret->shape()[0]);
+          dim_t seg_len = (ret->shape()[0] + num_threads - 1) / num_threads;
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: Should we replace `ret->shape()[0] ` by `nnr` instead? 
https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/indexing_op.h#L661

**Comment by eric-haibin-lin** [comment]
Yes, the original dot(csr.T, dense) performance is not good. The same for scipy - it's not always faster than the dense dot even with very sparse data.

File: src/operator/tensor/dot-inl.h (hunk 2 above)
Comment (ZiyueHuang) [reply src/operator/tensor/dot-inl.h]: I think no, this range should cover num_cols of csr, i.e. num_rows of output. So nnr is not helpful for this kernel.

File: src/operator/tensor/dot-inl.h [hunk 3]
```diff
@@ -375,28 +375,32 @@ struct DotCsrTransDnsRspByRowBlocks {
    * \brief
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: The documentation should be updated

**Comment by ZiyueHuang** [comment]
Added benchmark for `n=2` case,

Before,

```
python dot.py --num-omp-threads 16
========================================================
  mxnet sparse dot benchmark: dot(csr, default) = default
  (matrix multiplication: (m x k)^T * (k x n) = m x n)
========================================================
 lhs_density(%)  rhs_density(%)    context        m        k        n  t_sparse(ms)   t_dense(ms)  speedup
            1.0           100.0     cpu(0)      256  1000000        2         41.61         36.01     0.87
```

After,

```
python dot.py --num-omp-threads 16
========================================================
  mxnet sparse dot benchmark: dot(csr, default) = default
  (matrix multiplication: (m x k)^T * (k x n) = m x n)
========================================================
 lhs_density(%)  rhs_density(%)    context        m        k        n  t_sparse(ms)   t_dense(ms)  speedup
            1.0           100.0     cpu(0)      256  1000000        2         14.44         32.46     2.25
```

**Comment by moveforever** [comment]
hi, @eric-haibin-lin @ZiyueHuang , can the storage type of dot(csr.T, rsp) be rsp?

**Comment by eric-haibin-lin** [comment]
It's now indeed rsp:
```
a = mx.nd.ones((2,2)).tostype('row_sparse')
>>> a

<RowSparseNDArray 2x2 @cpu(0)>
>>> b = mx.nd.ones((2,2)).tostype('csr')
>>> mx.nd.sparse.dot(b,a,transpose_a=True)

<RowSparseNDArray 2x2 @cpu(0)>
```

**Comment by moveforever** [comment]
oh, i know, thanks.

**Pull request merged**